### PR TITLE
Add dispatcher package

### PR DESCRIPTION
**Pull request by alexcarol**

### CODE REVIEW

File: dispatcher/dispatcher_test.go [hunk 1]
```diff
@@ -0,0 +1,221 @@
+package dispatcher
```
Comment (gonzaloserrano) [inline dispatcher/dispatcher_test.go]: would be nice to have this in a separate package `dispatcher_test`

Comment (gonzaloserrano) [reply dispatcher/dispatcher_test.go]: also these tests needs some love

File: dispatcher/dispatcher_test.go [hunk 2]
```diff
@@ -0,0 +1,221 @@
...
+}
+
+type testInterface2 interface {
+	TestFunc2()
+}
+
+type testEvent1 struct {
+}
+
+type testEvent2 struct {
```
Comment (gonzaloserrano) [inline dispatcher/dispatcher_test.go]: should be called *Spy, see https://8thlight.com/blog/uncle-bob/2014/05/14/TheLittleMocker.html

File: dispatcher/dispatcher_test.go [hunk 3]
```diff
@@ -0,0 +1,221 @@
...
+	triggered bool
```
Comment (gonzaloserrano) [inline dispatcher/dispatcher_test.go]: a struct with just one field can be converted to `type testEvent2 bool`

File: dispatcher/dispatcher_test.go [hunk 4]
```diff
@@ -0,0 +1,221 @@
+package dispatcher
+
+import (
+	"testing"
+	"time"
+
+	"github.com/stretchr/testify/assert"
+)
+
+type testInterface1 interface {
```
Comment (gonzaloserrano) [inline dispatcher/dispatcher_test.go]: I don't like this type name, does not say anything

File: dispatcher/dispatcher_test.go [hunk 5]
```diff
@@ -0,0 +1,221 @@
...
+}
+
+type testEvent1 struct {
+}
+
+type testEvent2 struct {
+	triggered bool
+}
+
+func (testEvent1) TestFunc1()  {}
```
Comment (gonzaloserrano) [inline dispatcher/dispatcher_test.go]: methods of a type are put close to its type not somewhere else

File: dispatcher/dispatcher_test.go [hunk 6]
```diff
@@ -0,0 +1,221 @@
...
+func (*testEvent2) TestFunc2() {}
+
+func TestStaticListener(t *testing.T) {
+	assert := assert.New(t)
+
+	dispatcher := New()
+
+	triggered := false
```
Comment (gonzaloserrano) [inline dispatcher/dispatcher_test.go]: use zero value so `var triggered bool`

File: dispatcher/dispatcher_test.go [hunk 7]
```diff
@@ -0,0 +1,221 @@
...
+}
+
+type testEvent2 struct {
+	triggered bool
+}
+
+func (testEvent1) TestFunc1()  {}
+func (*testEvent2) TestFunc2() {}
+
+func TestStaticListener(t *testing.T) {
```
Comment (gonzaloserrano) [inline dispatcher/dispatcher_test.go]: what does "static" mean? use better naming

File: dispatcher/dispatcher_test.go [hunk 8]
```diff
@@ -0,0 +1,221 @@
...
+	listeners := []interface{}{
+		func(testEvent1) { triggered1 = true },
+		func(testEvent1) { triggered2 = true },
+	}
+	err := dispatcher.AddListener(listeners...)
+	assert.NoError(err)
+
+	dispatcher.Dispatch(testEvent1{})
+	assert.True(triggered1 && triggered2)
+}
```
Comment (gonzaloserrano) [inline dispatcher/dispatcher_test.go]: I'm quite confident all these previous tests can be combined in a single one with test cases, what do you think?

File: dispatcher/dispatcher_test.go [hunk 9]
```diff
@@ -0,0 +1,221 @@
...
+
+func TestBadListenerWrongType(t *testing.T) {
+	assert := assert.New(t)
+	dispatcher := New()
+
+	err := dispatcher.AddListener("this is not a function")
+	assert.Error(err)
+	assert.NotEmpty(err.Error())
+}
```
Comment (gonzaloserrano) [inline dispatcher/dispatcher_test.go]: this one can be combied with the prev one

File: dispatcher/dispatcher_test.go [hunk 10]
```diff
@@ -0,0 +1,221 @@
...
+
+func TestAsynchronousDispatch(t *testing.T) {
```
Comment (gonzaloserrano) [inline dispatcher/dispatcher_test.go]: I would remove this test, this does not show test anything but channels (?)

File: dispatcher/dispatcher_test.go [hunk 11]
```diff
@@ -0,0 +1,221 @@
...
+	triggered := false
+	dispatcher := New()
+
+	err := dispatcher.AddListener(func(testInterface2) { triggered = true })
+	assert.NoError(err)
+
+	dispatcher.Dispatch(&testEvent2{})
+
+	assert.True(triggered, "interface listener failed to trigger for pointer")
+}
```
Comment (gonzaloserrano) [inline dispatcher/dispatcher_test.go]: these 3 prev tests can be combined

**Comment by gonzaloserrano** [comment]
the code is good but the tests are too verbose IMHO

**Comment by alexcarol** [comment]
Thanks for the review @gonzaloserrano, I'll do a PR with some improvements next week with some improvements. About the readme, I think I'll probably move it to a doc.go file, so it appears in the godoc page.

**Pull request merged**